File: src/main/java/org/example/easy/AddBinary.java
package org.example.easy;

public class AddBinary {
    public String addBinary(String a, String b) {
        String result = addBinaryHelper(a, b, a.length() - 1, b.length() - 1, 0, "");
        return new StringBuilder(result).reverse().toString();
    }

    private String addBinaryHelper(String a, String b, int aLength, int bLength, int carry, String result) {
        if (aLength < 0 && bLength < 0) {
            return carry != 0 ? result + carry : result;
        }

        if (aLength >= 0) {
            carry += a.charAt(aLength) - '0';
        }

        if (bLength >= 0) {
            carry += b.charAt(bLength) - '0';
        }

        return addBinaryHelper(a, b, --aLength, --bLength, carry / 2, result + carry % 2);
    }
}

File: leetcode-python/src/solution/easy/may/add_binary.py
def add_binary(a: str, b: str) -> str:
    res = ''
    a_len = len(a) - 1
    b_len = len(b) - 1
    carry = 0
    while a_len >= 0 or b_len >= 0 or carry > 0:
        if a_len >= 0:
            carry += int(a[a_len])
            a_len -= 1
        if b_len >= 0:
            carry += int(b[b_len])
            b_len -= 1

        res = str(carry % 2) + res
        carry //= 2

    return res
